sui_programmability/framework/sources/Governance/ValidatorSet.move

```diff
@@ -0,0 +1,299 @@
+        Validator::request_withdraw_stake(validator, withdraw_amount, min_validator_stake);
+    }
+
+    /// Update the validator set at the end of epoch.
+    /// It does the following things:
+    ///   1. Distribute stake award.
+    ///   2. Process pending stake deposits and withdraws for each validator (`adjust_stake`).
+    ///   3. Process pending validator application and withdraws.
+    ///   4. At the end, we calculate the total stake for the new epoch.
+    public(friend) fun advance_epoch(
+        self: &mut ValidatorSet,
+        ctx: &mut TxContext,
+    ) {
+        // TODO: Distribute stake rewards.
+
+        // This needs to come before adjust_stake since some of the
+        // pending validators may also have pending deposits.
+        process_pending_validators(&mut self.active_validators, &mut self.pending_validators);
+
+        adjust_stake(&mut self.active_validators, ctx);
+
+        process_pending_removals(&mut self.active_validators, &mut self.pending_removals);
+
+        self.total_stake = calculate_total_stake(&self.active_validators);
+        self.quorum_threshold_pct = calculate_quorum_threshold(&self.active_validators);
+    }
+
+    /// Checks whether a duplicate of `new_validator` is already in `validators`.
+    /// Two validators duplicate if they share the same sui_address or same IP or same name.
+    fun contains_duplicate_validator(validators: &vector<Validator>, new_validator: &Validator): bool {
+        let len = Vector::length(validators);
+        let i = 0;
+        while (i < len) {
+            let v = Vector::borrow(validators, i);
+            if (Validator::is_duplicate(v, new_validator)) {
+                return true
+            };
+            i = i + 1;
+        };
+        false
+    }
+
+    /// Find validator by `validator_address`, in `validators`.
+    /// Returns (true, index) if the validator is found, and the index is its index in the list.
+    /// If not found, returns (false, 0).
+    fun find_validator(validators: &vector<Validator>, validator_address: address): Option<u64> {
+        let length = Vector::length(validators);
+        let i = 0;
+        while (i < length) {
+            let v = Vector::borrow(validators, i);
+            if (Validator::get_sui_address(v) == validator_address) {
+                return Option::some(i)
+            };
+            i = i + 1;
+        };
+        Option::none()
+    }
+
+    /// Process the pending withdraw requests. For each pending request, the validator
+    /// is removed from `validators` and sent back to the address of the validator.
+    fun process_pending_removals(validators: &mut vector<Validator>, withdraw_list: &mut vector<u64>) {
+        sort_removal_list(withdraw_list);
+        while (!Vector::is_empty(withdraw_list)) {
+            let index = Vector::pop_back(withdraw_list);
+            let validator = Vector::remove(validators, index);
```

Comment:
This will be O(n^2), but not sure there is much else we can do about it now

Reply:
hmm why is this O N^2?

Reply:
(assuming it is stable, which I think it is)

Reply:
Also is this even correct then? The indexes will all shift around after the first remove

Reply:
`sort_removal_list` will sort the indexes, so we will be removing the highest index first, and etc.

Reply:
Ah, gotcha.... Still O(n^2) worst case, but should be a bit better in practice (probably roughly O(n), especially if n is smallish)